include/uapi/sound/sof/header.h

```diff
@@ -26,6 +26,24 @@ struct sof_abi_hdr {
 	__u32 data[];		/**< Component data - opaque to core */
 }  __packed;
 
+/*
+ * Identifies data type, size and ABI for IPC4.
+ * Used by any bespoke component data structures or binary blobs.
+ */
+struct sof_ipc4_abi_hdr {
+	__u32 magic;		/**< 'S', 'O', 'F', '4' */
+	__u32 size;		/**< size in bytes of data excl. this struct */
+	__u32 abi;		/**< SOF ABI version */
+	__u32 blob_type;	/**< type of blob: MODULE_INIT, CONFIG or LARGE_CONFIG */
```

Comment:
where are the values for MODULE_INIT, CONFIG and LARGE_CONFIG defined?

A grep in the linux tree gave me nothing. I do see this

````c
enum sof_ipc4_module_type {
	SOF_IPC4_MOD_INIT_INSTANCE,
	SOF_IPC4_MOD_CONFIG_GET,
	SOF_IPC4_MOD_CONFIG_SET,
	SOF_IPC4_MOD_LARGE_CONFIG_GET,
	SOF_IPC4_MOD_LARGE_CONFIG_SET,
	SOF_IPC4_MOD_BIND,
	SOF_IPC4_MOD_UNBIND,
	SOF_IPC4_MOD_SET_DX,
	SOF_IPC4_MOD_SET_D0IX,
	SOF_IPC4_MOD_ENTER_MODULE_RESTORE,
	SOF_IPC4_MOD_EXIT_MODULE_RESTORE,
	SOF_IPC4_MOD_DELETE_INSTANCE,

	SOF_IPC4_MOD_TYPE_LAST,
};
````
but was it the intended structure? There no such thing as MODULE_INIT.

Also is this for read and write, or write only?

Reply:
Yes, the value comes from ```sof_ipc4_module_type```. I explained it in the commit message. Copying the whole name , such as SOF_IPC4_MOD_INIT_INSTANCE, causes to excess100 bytes limitation. I will update to `/**< type of blob: INIT_INSTANCE, CONFIG_SET or LARGE_CONFIG_SET */`

The INIT_INSTANCE should be read only. Topologies will set it and user space tools doesn't need to set it.

Reply:
why do we need to make it reead-only. In the end its just data isnt it? Topology data will set the blob_type to INIT_DATA. But sof-clt can use the same kcontrol to write data to the module with CONFIG_SET/LARGE_CONFIG. I see no need to separate kcontrols at all.

Reply:
> Yes, the value comes from `sof_ipc4_module_type`. I explained it in the commit message. Copying the whole name , such as SOF_IPC4_MOD_INIT_INSTANCE, causes to excess100 bytes limitation. I will update to `/**< type of blob: INIT_INSTANCE, CONFIG_SET or LARGE_CONFIG_SET */`
> 
> The INIT_INSTANCE should be read only. Topologies will set it and user space tools doesn't need to set it.

Please add it to the comment for the field. Maybe use kernek-style struct definition like we have in ipc4-topology.h